### PR TITLE
Has Binary Targets Flag

### PR DESCRIPTION
**Pull request by samdeane**
Implements #1025.

Work in progress right now.

- Adds `hasBinaryTargets` to the PackageShow model.
- Adds `binaryTargetsItem()` to the PackageShow metadata.
- Adds a `type` property to `Manifest.Target` to read the target type from SPM.
- Adds a `hasBinaryTargets` field to `Version`, which is set to true during analysis, if the manifest contains at least one binary target

### CODE REVIEW

**Comment by samdeane** [comment]
A few outstanding questions:

- I picked `binary` for the css class name, but `binaryOnly` or `binary-only` might be more appropriate, depending on your naming conventions
- The layout style is wrong and there's no icon, but I'm guessing that will need fixing in CSS?
- I wasn't sure if the README should link to actual README file, or just to the repo's main URL. If it's the latter, then I can remove the `readmeURL` property I added to the model.

**Comment by samdeane** [comment]
Just noticed this in your original comment: 

> With [README] being a link to the #readme anchor and [LICENSE] being a link to the license file (which we already have in the page).

So that answers the question about the `readmeURL`; I will remove that.

**Comment by samdeane** [comment]
Think I've figured out the css change, and have added the relevant bit. Do I need to run something to update the frontend?

**Review by daveverwer — COMMENTED** [review]
This is looking great so far, thank you so much @samdeane! I left a few inline comments, but nothing major at all!

**Comment by daveverwer** [comment]
> I re-ran ` make serve-front-end` which seemed to do the trick

You should only need to run this one, but it should stay running and watch for changes. My workflow is to run this command when I start working on anything CSS/JS-based but once it's running it should all be instant and automatic. Did it not work that way for you?

**Comment by finestructure** [comment]
Thanks for putting up this draft PR, Sam, and apologies for the late reply. I haven't had a chance to chime in sooner, except for the brief note earlier 😅

I would advise against joining with the targets table for a number of reasons. One is performance - it's certain to have an impact given it's our second highest cardinality table after builds. Another is complexity. You've already encountered the package page query and it's not the easiest query to extend. A third reason is that we not only pay the cost of the join but also a cost when deriving the aggregate into a boolean off of that join when querying.

However, I also think it's unnecessary to keep around detailed binary target information. As far as I can see in the package show model, the interest is a simple package level boolean "has binary targets". (Note: is this just _any_ binary targets? Is that useful by itself? Should it not be "has _only_ binary targets"?) *)

This feels like it is package level information: we're not querying anything target or even version specific. Unlike some other metadata like for instance build information (builds can and often do signal different compatibility between versions), binary target information is unlikely to change between versions. I'd argue it'd be something to track alongside license information (which it relates to) in `repositories` and therefore at the package level. This would make implementation much, much simpler.

One thing we should do, though, is assess the binary target landscape by grepping through all our checkouts on staging, as I suggested above. (That's something Dave or I need to do – I can take a look as soon as I get a chance.) This will tell us how many packages there are and what they look like. It might for instance inform a decision whether to stick this on `versions` or `repositories`. If there's only a handful, I'd definitely just stick this on `repositories` - if there's even a debate about that.

Hope that helps and thanks again for the PR!

*) I've now seen the discussion above where you address this. My comments was based off looking at the PR changes, so perhaps there are still some variables in the package show model that need adjusting to clarify?

**Comment by samdeane** [comment]
Thanks for the update @finestructure.

I get what you're saying, but I think you and @daveverwer are pointing me in slightly different directions :). 

That's totally fine of course, but choosing the approach is probably above my pay grade in this situation so I'm probably best to let you thrash it out.

**Comment by daveverwer** [comment]
Apologies for that @samdeane!

I don't think it's worth the effort of checking how many packages this will affect. It's going to be a very small number, but that’s what makes this feature worth implementing!

Having a binary-only target is rare and it’s easy to assume that all packages listed on a public index are open-source (or at least source-available). The value of this feature is increased *because* it only affects a few packages. I’d guess probably there are probably around 10 in the whole index that this will show for.

> However, I also think it's unnecessary to keep around detailed binary target information.

I was thinking it wouldn’t be limited to binary targets, but all targets. I want to include more detail on products and targets in the future, and I thought this might be a way to get started on that work rather than implement this as a boolean, and then roll that back later. 

However, it’s often wishful thinking and it does add more work now, so let’s do it your way and store the boolean. If/when we ever expand targets/products, we can convert this to use richer data structures later.

>  If there's only a handful, I'd definitely just stick this on repositories - if there's even a debate about that.

I don’t think this information fits that well as package-level information on `packages` or `repositories`. Right now we have a clear delineation between information derived from the package manifest and other versioned data like the checked-out git repository (`versions`) populated via analysis, and information fetched from online and unversioned sources (`repositories`) populated via ingestion. Even though this flag is unlikely to change between versions, I think we should keep this delineation where we can. 

On the front-end, we already pull the default branch version in so it will have no implications on the front-end implementation. On the back end, we already put everything from the package manifest into `versions`, so deriving this data and adding it there would be easier than needing to also update repositories at that time.  It also keeps our database easy to reason about, although it’s probably only Sven and I who understand the delineation right now! 😅

So, I’d suggest this as a combination of our two approaches:

- No changes to the `targets` table
- Adding one boolean to `versions` named `has_binary_targets` that is populated during analysis alongside other information from the package manifest.
- pulling that boolean off the `defaultBranchVersion` into the view model for display on the package page.

Does that sound OK to you, @finestructure?

**Comment by finestructure** [comment]
> Does that sound OK to you, @finestructure?

Yep, my main concern was the join on the `targets` table!

**Comment by samdeane** [comment]
Thanks guys.

I’ve been temporarily distracted by other work but will get back to it eventually.

**Comment by daveverwer** [comment]
> I’ve been temporarily distracted by other work but will get back to it eventually.

No rush!

File: Tests/AppTests/PackageShowModelTests.swift [hunk 1]
```diff
@@ -72,6 +72,24 @@ class PackageShowModelTests: SnapshotTestCase {
         assertSnapshot(matching: renderedHistory, as: .lines)
     }
 
+    func test_binary_only() throws {
```
[the remark this answers is not in the document]
Comment (samdeane) [reply Tests/AppTests/PackageShowModelTests.swift]: Have now also added `test_PackageShowView_binary_targets()` to `WebpageSnapshotTests`.

**Comment by samdeane** [comment]
I’m getting fairly close I think.

One question about the new field and/or migration.

I either need to force the analysis phase to re-run for all packages, in order to set a value for the new `has_binary_targets` property in the `Version` model, or I need to make it an optional value, and default to false if it’s not set.

What’s the normal approach in this situation?

**Comment by daveverwer** [comment]
> What’s the normal approach in this situation?

We will run a full re-analyze, but the code will need to be running on the servers with the migration applied *before* that re-analyze can start, so must be able to cope with NULL values in that field.

However, you could achieve this in a different way and keep the Swift code non-optional. If the migration made the field NON NULL and also gave a default value in the migration, it would work the same way but keep the Swift code simpler.

**Comment by daveverwer** [comment]
This is just excellent work, thank you so much @samdeane!

Once you're happy with your side of it we'll move it onto a repository branch so I can push the CSS and potential new icon (haven't decided if it needs one yet!), and we'll do the review there, too. No need to do that until you're happy though as we'll lose all this context.

I wish there was a way we could get CI to run for selected contributors when the branch is from an external fork.

**Comment by samdeane** [comment]
> If the migration made the field NON NULL and also gave a default value in the migration, it would work the same way but keep the Swift code simpler

Ah, I wondered if something like that would work! I’ve updated the migration with a default value. Not sure if I need to do anything else there to explicitly make it non-null, like adding `.required`?

> Once you're happy with your side of it we'll move it onto a repository branch

Go for it. I did wonder whether I was supposed to do that right from the off, but I didn’t know the etiquette and figured I’d cause less damage in a fork if I got it all wrong :)

**Comment by daveverwer** [comment]
> > If the migration made the field NON NULL and also gave a default value in the migration, it would work the same way but keep the Swift code simpler
> 
> Ah, I wondered if something like that would work! I’ve updated the migration with a default value. Not sure if I need to do anything else there to explicitly make it non-null, like adding `.required`?

Yes, add a required. Here's an example from the `products` definition:

```swift
.field("name", .string, .required)
```

> > Once you're happy with your side of it we'll move it onto a repository branch
> 
> Go for it. I did wonder whether I was supposed to do that right from the off, but I didn’t know the etiquette and figured I’d cause less damage in a fork if I got it all wrong :)

Whenever you're happy to move to review, create the branch. You have permission on this repo.

**Comment by samdeane** [comment]
Ok, I've moved it to a branch in the main repo and created a new PR #1920.

Will close this one.

**Pull request closed without merging**